lib/srv/termmanager.go

```diff
@@ -0,0 +1,247 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package srv
+
+import (
+	"io"
+	"sync"
+	"sync/atomic"
+
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+const maxHistory = 1000
+
+// TermManager handles the streams of terminal-like sessions.
+// It performs a number of tasks including:
+// - multiplexing
+// - history scrollback for new clients
+// - stream breaking
+type TermManager struct {
+	mu           sync.Mutex
+	writers      map[string]io.Writer
+	readerState  map[string]*int32
+	OnWriteError func(idString string, err error)
+	countWritten uint64
+	countRead    uint64
+	// buffer is used to buffer writes when turned off
+	buffer []byte
+	on     bool
+	// history is used to store the scrollback history sent to new clients
+	history []byte
+	// incoming is a stream of incoming stdin data
+	incoming chan []byte
+	// remaining is a partially read chunk of stdin data
+	// we only support one concurrent reader so this isn't mutex protected
+	remaining       []byte
+	readStateUpdate *sync.Cond
+	closed          *int32
+}
+
+// NewTermManager creates a new TermManager.
+func NewTermManager() *TermManager {
+	return &TermManager{
+		writers:         make(map[string]io.Writer),
+		readerState:     make(map[string]*int32),
+		closed:          new(int32),
+		readStateUpdate: sync.NewCond(&sync.Mutex{}),
+		incoming:        make(chan []byte, 100),
+	}
+}
+
+func (g *TermManager) writeToClients(p []byte) int {
+	g.history = append(g.history, p...)
+
+	var truncate int
+	if len(g.history) > maxHistory {
+		truncate = maxHistory
+	} else {
+		truncate = len(g.history)
+	}
+
+	g.history = g.history[:truncate]
+	atomic.AddUint64(&g.countWritten, uint64(len(p)))
+
+	for key, w := range g.writers {
+		_, err := w.Write(p)
+		if err != nil {
+			if err != io.EOF {
+				log.Warnf("Failed to write to remote terminal: %v", err)
+			}
+
+			if g.OnWriteError != nil {
+				g.OnWriteError(key, err)
+			}
+
+			delete(g.writers, key)
+		}
+	}
+
+	return len(p)
+}
+
+func (g *TermManager) Write(p []byte) (int, error) {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	if g.on {
+		g.writeToClients(p)
+	} else {
+		g.buffer = append(g.buffer, p...)
+	}
+
+	return len(p), nil
+}
+
+func (g *TermManager) Read(p []byte) (int, error) {
+	if len(g.remaining) > 0 {
+		n := copy(p, g.remaining)
+		g.remaining = g.remaining[n:]
+		return n, nil
+	}
+
+	q := make(chan struct{})
+	c := make(chan bool)
+	go func() {
+		g.readStateUpdate.L.Lock()
+		defer g.readStateUpdate.L.Unlock()
+
+		for {
+			g.mu.Lock()
+			on := g.on
+			g.mu.Unlock()
+
+			select {
+			case c <- on:
```

Comment:
The value of `g.on` could change between the time you check it and write to `c` - is that okay here?

Reply:
In fact, `g.on` and `c` look a lot like a `sync.Cond` - maybe a refactor is possible here.

Reply:
1. Yes, that's okay here.
2. I'd like to use sync.Cond directly here instead of via a proxy notifier channel but there is a good reason I haven't. I initially attempted that way but I couldn't figure out a way to get it to play nicely since sync.Cond isn't selectable. For that reason it's impossible to have one goroutine select between an incoming read message on `g.incoming` and a `Cond.Broadcast` updating the state to false. This makes it so that the new status won't be detecter until after a read which isn't desirable. We need to switch off immediately without processing any further data and that requires a select.

Reply:
The goroutine updating `c` does use sync.Cond for detecting those updates.

Reply:
Yeah, the only way to make it "selectable" is to `Broadcast` on multiple conditions (so have another routine that reads from `g.incoming` and broadcasts when a message is received. Then when you wake up you check both conditions and determine how to proceed.

I think what you have here is fine for now though.